packages/backend/src/modules/finality/analyzers/StarknetFinalityAnalyzer.ts

```diff
@@ -0,0 +1,80 @@
+import { assert } from '@l2beat/backend-tools'
+import {
+  ProjectId,
+  TrackedTxsConfigSubtype,
+  UnixTime,
+} from '@l2beat/shared-pure'
+import { BigNumber, utils } from 'ethers'
+import { z } from 'zod'
+
+import { RpcClient } from '../../../peripherals/rpcclient/RpcClient'
+import { StarknetClient } from '../../../peripherals/starknet/StarknetClient'
+import { LivenessRepository } from '../../tracked-txs/modules/liveness/repositories/LivenessRepository'
+import { BaseAnalyzer } from './types/BaseAnalyzer'
+
+const ZBigNumber = z.instanceof(BigNumber).transform(Number)
+
+const StarknetStateUpdate = z.object({
+  programOutput: z.array(ZBigNumber),
+  kzgProof: z.string(),
+})
+type StarknetStateUpdate = z.infer<typeof StarknetStateUpdate>
+
+export class StarknetFinalityAnalyzer extends BaseAnalyzer {
+  constructor(
+    protected override readonly provider: RpcClient,
+    protected override readonly livenessRepository: LivenessRepository,
+    protected override readonly projectId: ProjectId,
+    private readonly l2Provider: StarknetClient,
+  ) {
+    super(provider, livenessRepository, projectId)
+  }
+  override getTrackedTxSubtype(): TrackedTxsConfigSubtype {
+    return 'stateUpdates'
+  }
+
+  async getFinality(transaction: {
+    txHash: string
+    timestamp: UnixTime
+  }): Promise<number[]> {
+    const tx = await this.provider.getTransaction(transaction.txHash)
+    const l1Timestamp = transaction.timestamp
+
+    const decodedTransactionData = decodeTransaction(tx.data)
+
+    const l2BlockNumber = extractBlockNumber(decodedTransactionData)
+
+    const { timestamp: l2Timestamp } = await this.l2Provider.getBlock(
+      l2BlockNumber,
+    )
+
+    return [l1Timestamp.toNumber() - l2Timestamp]
+  }
+}
+
+function decodeTransaction(data: string) {
+  const signature = `updateStateKzgDA(uint256[] programOutput, bytes kzgProof)`
+  const iface = new utils.Interface([`function ${signature}`])
+
+  const decodedInput = iface.decodeFunctionData(signature, data)
+
+  return StarknetStateUpdate.parse({
+    /* eslint-disable @typescript-eslint/no-unsafe-assignment */
+    programOutput: decodedInput.programOutput,
+    kzgProof: decodedInput.kzgProof,
+    /* eslint-enable @typescript-eslint/no-unsafe-assignment */
+  })
```

Comment:
Can't you just parse the decoded?
```suggestion
  return StarknetStateUpdate.parse(decodedInput)
```

Reply:
Ethers when parsing is doing some weird mumbo jumbo and instead of named properties also provides you with something like this if you are using precise iface:
```ts
[
  "foo",
  "bar",
  "foo": "foo",
  "foo": "foo",
]
```